File: android/src/main/java/expo/modules/backgroundremoval/ExpoBackgroundRemovalModule.kt
package expo.modules.backgroundremoval

import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ProcessOptions : Record {
  @Field
  var maxDimension: Int? = null
}

class RemoveBackgroundOptions : Record {
  @Field
  var cropToSubject: Boolean = false

  @Field
  var maxDimension: Int? = null
}

class ExpoBackgroundRemovalModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBackgroundRemoval")

    // Coroutine bodies run on Expo's shared modules queue, so the heavy work moves to
    // Dispatchers.Default rather than holding that queue for every other module.
    AsyncFunction("removeBackground") Coroutine { uri: String, options: RemoveBackgroundOptions ->
      return@Coroutine withContext(Dispatchers.Default) {
        processor().removeBackground(uri, options.cropToSubject, options.maxDimension)
      }
    }

    AsyncFunction("toPng") Coroutine { uri: String, options: ProcessOptions ->
      return@Coroutine withContext(Dispatchers.Default) {
        processor().toPng(uri, options.maxDimension)
      }
    }

    AsyncFunction("segmentImage") Coroutine { uri: String, options: ProcessOptions ->
      return@Coroutine withContext(Dispatchers.Default) {
        processor().segmentImage(uri, options.maxDimension)
      }
    }

    AsyncFunction("extractObjects") Coroutine { uri: String, options: ProcessOptions ->
      return@Coroutine withContext(Dispatchers.Default) {
        processor().extractObjects(uri, options.maxDimension)
      }
    }

    // Present for API parity: on Android it only renders the image, see the view's docs.
    View(ExpoBackgroundRemovalView::class) {
      Events("onSubjects", "onLoad")

      Prop("source") { view: ExpoBackgroundRemovalView, uri: String? ->
        view.setSource(uri)
      }

      Prop("highlightSubjects") { view: ExpoBackgroundRemovalView, highlight: Boolean ->
        view.setHighlightSubjects(highlight)
      }
    }
  }

  private fun processor(): BackgroundRemovalProcessor {
    val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
    return BackgroundRemovalProcessor(context, appContext.cacheDirectory)
  }
}

File: ios/ExpoBackgroundRemovalModule.swift
import ExpoModulesCore

internal struct ProcessOptions: Record {
  @Field var maxDimension: Int?
}

internal struct RemoveBackgroundOptions: Record {
  @Field var cropToSubject: Bool = false
  @Field var maxDimension: Int?
}

public class ExpoBackgroundRemovalModule: Module {
  // Expo's default async queue is shared by every module; keep the heavy work off it.
  private let processingQueue = DispatchQueue(label: "expo.modules.backgroundremoval", qos: .userInitiated)

  public func definition() -> ModuleDefinition {
    Name("ExpoBackgroundRemoval")

    AsyncFunction("removeBackground") { (uri: String, options: RemoveBackgroundOptions) -> [String: Any] in
      guard #available(iOS 17.0, *) else {
        throw Self.unsupportedOS()
      }
      return try Self.makeProcessor().removeBackground(
        uri: uri,
        cropToSubject: options.cropToSubject,
        maxDimension: options.maxDimension
      ).dictionary
    }
    .runOnQueue(processingQueue)

    AsyncFunction("toPng") { (uri: String, options: ProcessOptions) -> [String: Any] in
      guard #available(iOS 17.0, *) else {
        throw Self.unsupportedOS()
      }
      return try Self.makeProcessor().toPng(uri: uri, maxDimension: options.maxDimension).dictionary
    }
    .runOnQueue(processingQueue)

    AsyncFunction("segmentImage") { (uri: String, options: ProcessOptions) -> [String: Any] in
      guard #available(iOS 17.0, *) else {
        throw Self.unsupportedOS()
      }
      return try Self.makeProcessor().segmentImage(uri: uri, maxDimension: options.maxDimension)
    }
    .runOnQueue(processingQueue)

    AsyncFunction("extractObjects") { (uri: String, options: ProcessOptions) -> [[String: Any]] in
      guard #available(iOS 17.0, *) else {
        throw Self.unsupportedOS()
      }
      return try Self.makeProcessor().extractObjects(uri: uri, maxDimension: options.maxDimension)
    }
    .runOnQueue(processingQueue)

    // Apple's native subject lifting. iOS only; on Android the view just shows the image.
    View(ExpoBackgroundRemovalView.self) {
      Events("onSubjects", "onLoad")

      Prop("source") { (view: ExpoBackgroundRemovalView, uri: String?) in
        view.setSource(uri)
      }

      Prop("highlightSubjects") { (view: ExpoBackgroundRemovalView, highlight: Bool) in
        view.setHighlightSubjects(highlight)
      }
    }
  }

  // Static so the function bodies don't retain the module.
  @available(iOS 17.0, *)
  private static func makeProcessor() -> BackgroundRemovalProcessor {
    let cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
      ?? FileManager.default.temporaryDirectory
    return BackgroundRemovalProcessor(cacheDirectory: cacheDirectory)
  }

  private static func unsupportedOS() -> BackgroundRemovalException {
    return BackgroundRemovalException(
      .unsupportedOS,
      "Foreground instance segmentation requires iOS 17 or newer."
    )
  }
}
